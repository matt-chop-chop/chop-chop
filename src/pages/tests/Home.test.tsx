import { shallow } from "enzyme";
import React from "react";
import Home from "../index.page";

jest.mock("react-query", () => ({
  useQuery: jest.fn().mockReturnValue({
    data: {
      data: {
        meals: [
          { strIngredient: "Test 1", strDescription: "Test description" },
          { strIngredient: "Test 2", strDescription: "Test description" },
        ],
      },
    },
    isLoading: false,
  }),
}));

const useState = jest.spyOn(React, "useState");

beforeEach(() => {
  useState.mockClear();
});

describe("Home", () => {
  it("is exported from Home", () => {
    expect(Home).toBeDefined();
  });

  it("renders as expected", () => {
    const wrapper = shallow(<Home />);

    expect(wrapper).toMatchSnapshot();
  });

  it("calls setState when the ingredient FilterSelectForm setSelection is triggered", () => {
    const setFilter = jest.fn();
    useState.mockReturnValue(["None", setFilter]);

    const wrapper = shallow(<Home />);

    const FilterSelectForm = wrapper.find("FilterSelectForm");
    expect(FilterSelectForm).toHaveLength(3);

    const setSelection = FilterSelectForm.first().prop("setSelection") as (
      ingredient: string
    ) => void;
    setSelection("Tiger Prawns");
    expect(setFilter).toHaveBeenCalledWith("Tiger Prawns");
  });

  it("calls setState when the category FilterSelectForm setSelection is triggered", () => {
    const setFilter = jest.fn();
    useState.mockReturnValue(["None", setFilter]);

    const wrapper = shallow(<Home />);

    const FilterSelectForm = wrapper.find("FilterSelectForm");
    expect(FilterSelectForm).toHaveLength(3);

    const setSelection = FilterSelectForm.at(1).prop("setSelection") as (
      category: string
    ) => void;
    setSelection("Side");
    expect(setFilter).toHaveBeenCalledWith("Side");
  });

  it("calls setState when the area FilterSelectForm setSelection is triggered", () => {
    const setFilter = jest.fn();
    useState.mockReturnValue(["None", setFilter]);

    const wrapper = shallow(<Home />);

    const FilterSelectForm = wrapper.find("FilterSelectForm");
    expect(FilterSelectForm).toHaveLength(3);

    const setSelection = FilterSelectForm.last().prop("setSelection") as (
      area: string
    ) => void;
    setSelection("French");
    expect(setFilter).toHaveBeenCalledWith("French");
  });
});
