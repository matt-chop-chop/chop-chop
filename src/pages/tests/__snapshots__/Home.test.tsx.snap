// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders as expected 1`] = `
<Box>
  <Hero
    message="Chop Chop is a site designed to allow you to quickly and easily make restaurant quality dishes with ingredients you already have at home."
    title="Welcome to Chop Chop!"
  />
  <Flex
    flexDirection="column"
    mb="40px"
    mt="40px"
  >
    <Heading
      as="h2"
      variant="heading-medium"
    >
      What kind of recipe would you like to cook?
    </Heading>
    <FilterSelectForm
      errorMessage="Ingredients could not be loaded"
      id="select-ingredient"
      label="What would you like the main ingredient of your recipe to be?"
      loading={false}
      options={
        [
          {
            "description": "",
            "name": "None",
          },
          {
            "description": "Test description",
            "name": "Test 1",
          },
          {
            "description": "Test description",
            "name": "Test 2",
          },
        ]
      }
      setSelection={[Function]}
    />
    <FilterSelectForm
      errorMessage="Categories could not be loaded"
      id="select-category"
      label="What category of recipe would you like to cook?"
      loading={false}
      options={
        [
          {
            "name": "None",
          },
          {
            "name": "",
          },
          {
            "name": "",
          },
        ]
      }
      setSelection={[Function]}
    />
    <FilterSelectForm
      errorMessage="Areas could not be loaded"
      id="select-area"
      label="What area would you like your recipe to be from?"
      loading={false}
      options={
        [
          {
            "name": "None",
          },
          {
            "name": "",
          },
          {
            "name": "",
          },
        ]
      }
      setSelection={[Function]}
    />
  </Flex>
  <FilterResults
    area="None"
    category="None"
    ingredient="None"
  />
</Box>
`;
